lib/pg/connection.rb

```diff
@@ -47,7 +47,7 @@ def self::parse_connect_args( *args )
 
 		if args.length == 1
 			case args.first
-			when URI, /\A#{URI.regexp}\z/
+			when URI, /\A#{URI::ABS_URI}\z/
```

Comment:
I like this approach, but I want to point out that this constant isn't an exact replacement.

```
irb(main):001:0> RUBY_VERSION
=> "2.6.0"

irb(main):002:0> URI.regexp == URI::ABS_URI
(irb):2: warning: URI.regexp is obsolete
=> false

irb(main):003:0> URI.regexp == URI::ABS_URI_REF
(irb):3: warning: URI.regexp is obsolete
=> true
```

`ABS_URI` anchors to the beginning and end like we do here and also allows some whitespace:

```
irb(main):004:0> URI::ABS_URI
=> /\A\s*
...
      \s*\z/x
```

I wonder if using a `URI` regexp is overkill for this situation. Should we simply check for a prefix of `postgresql://` or `postgres://` as described in the [libpq docs](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING)?

Reply:
@cbandy thank you for your checking!

I will replace `URI::ABS_URI` with `URI::ABS_URI_REF`.

> I wonder if using a URI regexp is overkill for this situation. Should we simply check for a prefix of postgresql:// or postgres:// as described in the libpq docs?

I do not understand what is the possible values of `args.first` in the context well.
The check does not judge wrongly a value that is previously judged as `else` condition?

```
      case args.first
      when URI, /\A#{URI::ABS_URI}\z/
        uri = URI(args.first)
        options.merge!( Hash[URI.decode_www_form( uri.query )] ) if uri.query
      when /=/
        # Option string style
        option_string = args.first.to_s
      else
        # Positional parameters
        options[CONNECT_ARGUMENT_ORDER.first.to_sym] = args.first
      end
```

Reply:
> I do not understand what is the possible values of `args.first` in the context well.

I agree, this method is doing more than is even documented on [`Connection.new`](https://deveiate.org/code/pg/PG/Connection.html#method-c-new). The [tests](https://github.com/ged/ruby-pg/blob/master/spec/pg/connection_spec.rb#L12) give a more complete explanation.

This method exists to handle `Connection.new`'s ... [overloaded](https://en.wikipedia.org/wiki/Function_overloading) ... arguments. When there is only one argument, this tries to detect if it is an entire connection string or if it is just a hostname.

> The check does not judge wrongly a value that is previously judged as `else` condition?

I think I don't understand the question. The `else` portion of a `case` statement is executed _after_ all of the expressions in `when` portions are falsey.

http://ruby-doc.org/core/doc/syntax/control_expressions_rdoc.html#label-case+Expression

Reply:
> I think I don't understand the question. The else portion of a case statement is executed after all of the expressions in when portions are falsey.

I mean for example when a value of `args.first` is `postgresql://apostgresql://`, before this modification, the value is evaluated in the `else` statement. But after this modification with `postgresql://` or `postgres://` is okay for the 1st `when`, the value `postgresql://apostgresql://` is NOT evaluated in the `else` statement.

I was worry about this kind of case.

Reply:
Ah sorry, `postgresql://apostgresql://` matches both before and after this PR.

```
irb(main):014:0> /\A#{URI.regexp}\z/.match('postgresql://apostgresql://') != nil
=> true
```

I meant if there is a value that matches 1st `when` before this PR and does not match 1st `when` after this PR when we fix by "postgresql:// or postgres:// matches for 1st when". I thought that could be issue.

Reply:
Agreed. I think that may be possible, but I would be surprised if it is a problem in practice. (For example, I don't think DNS names are allowed to have slashes.)

If someone's `host` is `postgresql://noturl`, then I'm okay requiring that they call it with keyword (hash) arguments:

```ruby
PG.connect(host: 'postgresql://noturl')
```

I'll leave it to maintainers to decide!